docs/csharp/language-reference/keywords/partial-method.md

```diff
@@ -14,9 +14,15 @@ A partial method has its signature defined in one part of a partial type, and it
 
 - Signatures in both parts of the partial type must match.
 
-- The method must return void.
+Partial method must have an implementation part in following cases:
 
-- No access modifiers are allowed. Partial methods are implicitly private.
+- It has any accessibility modifiers (even [private](../../language-reference/keywords/private.md)).
+
+- It has any return type other than [void](../../language-reference/builtin-types/void.md).
+
+- It has [out](../../language-reference/keywords/out-parameter-modifier.md) parameters.
+
+- It has a [virtual](../../language-reference/keywords/virtual.md), [override](../../language-reference/keywords/override.md), [sealed](../../language-reference/keywords/sealed.md), [new](../../language-reference/keywords/new-modifier.md), or [extern](../../language-reference/keywords/extern.md) modifier.
```

Comment:
Here, and in the following change noted, I think it would be easier to define this by stating when a `partial` method *doesn't* need to have an implementation.  Namely, a partial method that returns `void`, and doesn't include any modifiers (including the default `private` access modifier) doesn't require an implementation. 

I like transposing it this way for two reasons: First, it's a simpler definition. Second, it provides a clear mental model to how the new features don't represent a breaking change.

Reply:
A partial method doesn't need to have an implementation if it returns `void`, doesn't have any `out` parameters, doesn't include any modifiers (including the default private access modifier) and doesn't have `extern` modifier. 

So condition for "optional implementation" is also not that simple.

Reply:
@maxkoshevoi 

> So condition for "optional implementation" is also not that simple.

It's not that simple, but it is simpler than what's written. And, my second point still applies. Effectively, partial methods don't need an implementation if they follow the rules in place before C# 9.0

Reply:
@BillWagner Yes, I see your point, but don't quite get, how you want to phrase it.

Something like:

> Implementation part is optional, if all of the conditions are met:
>
> - Partial method doesn't have any accessibility modifiers (even [private](../../language-reference/keywords/private.md)).
> - Partial method returns [void](../../language-reference/builtin-types/void.md).
> - Partial method doesn't have any [out](../../language-reference/keywords/out-parameter-modifier.md) parameters.
> - Partial method doesn't have any of the following modifiers: [virtual](../../language-reference/keywords/virtual.md), [override](../../language-reference/keywords/override.md), [sealed](../../language-reference/keywords/sealed.md), [new](../../language-reference/keywords/new-modifier.md), or [extern](../../language-reference/keywords/extern.md).

Or do you want to duplicate conditions like this:

>  A partial method doesn't need to have an implementation if it returns void, doesn't have any out parameters, doesn't include any modifiers (including the default private access modifier) and doesn't have extern modifier.
>
> Partial method must have an implementation part in the following cases:
>
> It has any accessibility modifiers
> ...

Reply:
Thanks @maxkoshevoi.  I think a mix of the two reads best. Let me know your thoughts.  Something like:

> A `partial` method doesn't need to have an implementation if it returns `void`, doesn't have any `out` parameters, doesn't include any modifiers (including the default `private` access modifier) and doesn't have the `extern` modifier.
>
> Any method that doesn't conform to all those restrictions (for example, `virtual public` methods), must provide an implementation.

Reply:
@BillWagner Looks good! I wanted to add those restrictions as a list for better readability. Updated PR. What do you think about correct version?

PS: Also added example of using partial method with source generator (since those changes were made mainly for that reason)